partitioned-heat-conduction/solver-fenics/heat_pySDC_problem_class.py

```diff
@@ -0,0 +1,142 @@
+            coupling_expr: FEniCS expression for the coupling boundary condition
+            precice_ref: preCICE-FEniCS adapter object reference
+            participant_name: Name of the participant (Dirichlet or Neumann)
+        """
+
+        # Set precice reference and coupling expression reference to update coupling boundary
+        # at every step within pySDC
+        self.precice = precice_ref
+        self.coupling_expression = coupling_expr
+        self.t_start = 0.0
+        self.participant_name = participant_name
+
+        # set mesh and function space for future reference
+        self.mesh = mesh
+        self.V = function_space
+
+        # invoke super init
+        super(fenics_heat_2d, self).__init__(self.V)
+
+        # Define Trial and Test function
+        u = TrialFunction(self.V)
+        v = TestFunction(self.V)
+
+        # Mass term
+        a_M = u * v * dx
+        self.M = assemble(a_M)
+
+        # Stiffness term (Laplace)
+        a_K = -1.0 * inner(nabla_grad(u), nabla_grad(v)) * dx
+        self.K = assemble(a_K)
+
+        # Forcing term
+        self.forcing_term_expr = forcing_term_expr
+
+        # Solution expression for error comparison and as boundary condition
+        # on the non-coupling boundary
+        self.solution_expr = solution_expr
+
+        # Currently only for Dirichlet boundary, has to be changed for Neumann boundary
+        if self.participant_name == 'Dirichlet':
+            self.couplingBC = DirichletBC(self.V, coupling_expr, coupling_boundary)
+
+        self.remainingBC = DirichletBC(self.V, solution_expr, remaining_boundary)
+
+        # Allow for fixing the boundary conditions for the residual computation
+        # Necessary if imex-1st-order-mass sweeper is used
+        self.fix_bc_for_residual = True
+
+        # define the homogeneous Dirichlet boundary for residual correction
+        def FullBoundary(x, on_boundary):
+            return on_boundary
+        self.homogenousBC = DirichletBC(self.V, Constant(0), FullBoundary)
+
+    def solve_system(self, rhs, factor, u0, t):
+        """
+        Solves (M - factor * A) u = rhs.
+        """
+        u = self.dtype_u(u0)
+        T = self.M - factor * self.K
+        b = self.dtype_u(rhs)
+
+        self.solution_expr.t = t
+        self.remainingBC.apply(T, b.values.vector())
+
+        # Update the coupling boundary condition for the Dirichlet participant
+        if self.participant_name == 'Dirichlet':
+            dt = t - self.t_start                   # This dt is used to read data from the current time window
+            read_data = self.precice.read_data(dt)  # Read the data to update the coupling expression
+            self.precice.update_coupling_expression(
+                self.coupling_expression,
+                read_data)    # Update the coupling expression
+
+            self.couplingBC.apply(T, b.values.vector())  # Apply the coupling boundary condition
```

Comment:
Assuming `self.dirichlet_bcs` is a list of all `DirichletBC`s. I don't think we need to differentiate between coupling and non-coupling BCs.

```suggestion
    for bc in self.dirichlet_bcs:
        bc.apply(T, b.values.vector())  # Apply all Dirichlet boundary conditions
```